libcloud/compute/drivers/vsphere.py

```diff
@@ -49,506 +34,206 @@
-            request.set_element__this(_this)
-            ret = server._proxy.UnregisterVM(request)
-            task = VITask(ret, server)
-
-        return True
+            context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+            context.verify_mode = ssl.CERT_NONE
 
-    @wrap_non_libcloud_exceptions
-    def ex_stop_node(self, node):
-        vm = self._get_vm_for_node(node=node)
-        vm.power_off()
+        try:
+            self.conn = connect.SmartConnect(host=host,
+                                             user=username,
+                                             pwd=password,
+                                             sslContext=context)
+            atexit.register(connect.Disconnect, self.conn)
+        except Exception as e:
+            err = str(e).lower()
+            if "incorrect user name" in err:
+                raise InvalidCredsError("username/password not acepted")
+            if "connection refused" in err or "is not a vim server" in err:
+                raise Exception("host doesn't appear to be a vSphere instance")
+            if "name or service not known" in err:
+                raise Exception("couldn't connect to '{}'".format(host))
+            if "certificate verify failed" in err:
+                raise Exception("SSL certificate couldn't be verified; if "
+                                "this is expected `verify_ssl=False` can "
+                                "be added to the driver to suppress checks")
+            raise Exception("unexpected error: {}".format(e))
+
+    @property
+    def _root(self):
+        return self.conn.RetrieveContent().rootFolder
+
+    @property
+    def _datacenters(self):
+        return [dc for dc in self._root.childEntity
+                if type(dc) is vim.Datacenter]
+
+    @property
+    def _vms(self):
+        vms = []
+        nodes = [n for dc in self._datacenters
+                 for n in dc.vmFolder.childEntity]
+
+        def vms_in(folder):
+            """ recursively extract the vms from a folder """
+            vms = []
+            for node in folder.childEntity:
+                if type(node) is vim.Folder:
+                    vms += vms_in(node)
+                elif type(node) is vim.VirtualMachine:
+                    vms.append(node)
+            return vms
+
+        for node in nodes:
+            if type(node) is vim.Folder:
+                vms += vms_in(node)
+            elif type(node) is vim.VirtualMachine:
+                vms.append(node)
+
+        return vms
+
+    def _get_obj_by_name(self, vimtype, name):
+        if type(vimtype) is not list:
+            vimtype = [vimtype]
+
+        content = self.conn.RetrieveContent()
+        container = content.viewManager.CreateContainerView(
+            content.rootFolder,
+            vimtype,
+            True)
 
-        return True
+        try:
+            return [i for i in container.view if i.name == name][0]
```

Comment:
This again computes the whole list to throw away most of it. Use a generator again:

```python
next(i for i in container.view if i.name == name)
```

Or combine `next` with `filter`